code/datums/components/supermatter_crystal.dm

```diff
@@ -0,0 +1,369 @@
+		if(!iseffect(consumed_object))
+			var/suspicion = ""
+			if(consumed_object.fingerprintslast)
+				suspicion = "last touched by [consumed_object.fingerprintslast]"
+				message_admins("[atom_source] has consumed [consumed_object], [suspicion] [ADMIN_JMP(atom_source)].")
+			atom_source.investigate_log("has consumed [consumed_object] - [suspicion].", INVESTIGATE_ENGINE)
+		qdel(consumed_object)
+	if(!iseffect(consumed_object) && isitem(consumed_object))
+		var/obj/item/consumed_item = consumed_object
+		object_size = consumed_item.w_class
+		matter_increase += 70 * object_size
+
+	//Some poor sod got eaten, go ahead and irradiate people nearby.
+	radiation_pulse(atom_source, max_range = 6, threshold = 1.2 / max(object_size, 1), chance = 10 * object_size)
+	for(var/mob/living/near_mob in range(10))
+		atom_source.investigate_log("has irradiated [key_name(near_mob)] after consuming [consumed_object].", INVESTIGATE_ENGINE)
+		if (HAS_TRAIT(near_mob, TRAIT_RADIMMUNE) || issilicon(near_mob))
+			continue
+		if(ishuman(near_mob) && SSradiation.wearing_rad_protected_clothing(near_mob))
+			continue
+		if(near_mob in view())
+			near_mob.show_message(span_danger("As \the [atom_source] slowly stops resonating, you find your skin covered in new radiation burns."), MSG_VISUAL,
+				span_danger("The unearthly ringing subsides and you find your skin covered in new radiation burns."), MSG_AUDIBLE)
+		else
+			near_mob.show_message(span_hear("An unearthly ringing fills your ears, and you find your skin covered in new radiation burns."), MSG_AUDIBLE)
+	consume_returns(matter_increase, damage_increase)
+
+/datum/component/supermatter_crystal/proc/consume_returns(matter_increase = 0, damage_increase = 0)
+	if(consume_callback)
+		consume_callback.Invoke(matter_increase, damage_increase)
+
+/datum/component/supermatter_crystal/proc/use_scalpel(obj/item/scalpel/supermatter/scalpel, mob/living/user, atom/atom_source)
+	to_chat(user, span_notice("You carefully begin to scrape \the [atom_source] with \the [scalpel]..."))
+	if(scalpel.use_tool(atom_source, user, 60, volume=100))
+		if (scalpel.usesLeft)
+			to_chat(user, span_danger("You extract a sliver from \the [atom_source]. \The [atom_source] begins to react violently!"))
+			new /obj/item/nuke_core/supermatter_sliver(atom_source.drop_location())
+			consume_returns(matter_increase = 800)
+			scalpel.usesLeft--
+			if (!scalpel.usesLeft)
+				to_chat(user, span_notice("A tiny piece of \the [scalpel] falls off, rendering it useless!"))
+		else
+			to_chat(user, span_warning("You fail to extract a sliver from \The [atom_source]! \the [scalpel] isn't sharp enough anymore."))
+
+/datum/component/supermatter_crystal/proc/use_destabilizing_crystal(obj/item/destabilizing_crystal, mob/living/user, atom/atom_source)
+	if(!istype(atom_source, /obj/machinery/power/supermatter_crystal))
+		to_chat(user, span_warning("You can only use \the [destabilizing_crystal] on a Full Supermatter Crystal."))
+		return
```

Comment:
Any reason why this is on the component itself instead of the actual supermatter crystal object? this istype is icky

Reply:
attack_by is done here, but i can move it to the object itself